### PR TITLE
Header row index support

### PR DESCRIPTION
**Pull request by waldimen**
I implemented `HeaderRowIndex` for import and export.

### CODE REVIEW

File: Npoi.Mapper/src/Npoi.Mapper/Mapper.cs [hunk 1]
```diff
@@ -503,12 +503,13 @@ public void Save<T>(Stream stream, int sheetIndex = 0, bool overwrite = true, bo
 
         private IEnumerable<RowInfo<T>> Take<T>(ISheet sheet, int maxErrorRows, Func<T> objectInitializer = null) where T : class
         {
+
```
Comment (donnytian) [inline Npoi.Mapper/src/Npoi.Mapper/Mapper.cs]: pls remove blank line

File: Npoi.Mapper/src/Npoi.Mapper/Mapper.cs [hunk 2]
```diff
@@ -503,12 +503,13 @@ public void Save<T>(Stream stream, int sheetIndex = 0, bool overwrite = true, bo
...
             if (sheet == null || sheet.PhysicalNumberOfRows < 1)
             {
                 yield break;
             }
 
-            var firstRowIndex = sheet.FirstRowNum;
+            var firstRowIndex = HeaderRowIndex == -1 ? sheet.FirstRowNum : HeaderRowIndex;
```
Comment (donnytian) [inline Npoi.Mapper/src/Npoi.Mapper/Mapper.cs]:
```cs
var firstRowIndex = HeaderRowIndex < 0 ? sheet.FirstRowNum : HeaderRowIndex;
                                   ^^^
```

File: Npoi.Mapper/src/Npoi.Mapper/Mapper.cs [hunk 3]
```diff
@@ -862,16 +865,17 @@ private void LoadWorkbookFromFile(string path)
         private void Put<T>(ISheet sheet, IEnumerable<T> objects, bool overwrite)
         {
             var sheetName = sheet.SheetName;
-            var firstRow = sheet.GetRow(sheet.FirstRowNum);
+            int firstRowIndex = HeaderRowIndex == -1 ? sheet.FirstRowNum : HeaderRowIndex;
```
Comment (donnytian) [inline Npoi.Mapper/src/Npoi.Mapper/Mapper.cs]: same here

File: Npoi.Mapper/src/Npoi.Mapper/Mapper.cs [hunk 4]
```diff
@@ -936,7 +940,8 @@ private void Save<T>(Stream stream, ISheet sheet, IEnumerable<T> objects, bool o
 
         private IRow PopulateFirstRow(ISheet sheet, List<ColumnInfo> columns, Type type)
         {
-            var row = sheet.CreateRow(sheet.FirstRowNum);
+            int firstRowIndex = HeaderRowIndex == -1 ? sheet.FirstRowNum : HeaderRowIndex;
```
Comment (donnytian) [inline Npoi.Mapper/src/Npoi.Mapper/Mapper.cs]: same here

File: Npoi.Mapper/test/ExportCustomResolverTests.cs [hunk 5]
```diff
@@ -41,5 +41,37 @@ public void ForHeader_ChangeHeaderStyle_ShouldChanged()
...
+                    cell.CellStyle = style;
+                }
+            });
+            mapper.Put(new[] { new SampleClass { StringProperty = str1 } });
+
+            // Assert
+            var row2 = workbook.GetSheetAt(0).GetRow(1 + mapper.HeaderRowIndex);
+            Assert.AreEqual(120, row1.GetCell(11).CellStyle.LeftBorderColor);
+            Assert.AreEqual(str1, row2.GetCell(11).StringCellValue);
+
```
Comment (donnytian) [inline Npoi.Mapper/test/ExportCustomResolverTests.cs]: pls remove blank line

File: Npoi.Mapper/test/AttributeTests.cs [hunk 6]
```diff
@@ -137,5 +137,29 @@ public void IgnoreAttributeTest()
             // Assert
             Assert.IsNull(objs[0].Value.IgnoredAttributeProperty);
         }
+
+        [TestMethod]
+        public void ColumnAttributeIndexTestFromGivenIndexHeader()
+        {
+            // Prepare
```
Comment (donnytian) [inline Npoi.Mapper/test/AttributeTests.cs]: Let's use "Arrange" for all new tests.

File: Npoi.Mapper/test/ExportGeneralTests.cs [hunk 7]
```diff
@@ -414,5 +414,86 @@ public void Map_WithIndexAndName_ShouldExportCustomColumnName()
...
+            exporter.Put(new[] { sampleObj }, "Resources");
+            var workbook = WriteAndReadBack(exporter.Workbook, existingFile);
+
+            // Assert
+            var sheet = workbook.GetSheet(sheetName);
+            Assert.AreEqual(sampleObj.GeneralProperty, sheet.GetRow(1 + headerRowIndex).GetCell(1).StringCellValue);
+            Assert.AreEqual(sampleObj.DateProperty.Date, sheet.GetRow(1 + headerRowIndex).GetCell(2).DateCellValue.Date);
+        }
+
+
```
Comment (donnytian) [inline Npoi.Mapper/test/ExportGeneralTests.cs]: blank lines

File: Npoi.Mapper/test/ImportGeneralTests.cs [hunk 8]
```diff
@@ -443,5 +443,46 @@ public void Map_WithIndexAndName_ShouldImportByIndex()
...
         }
+
+        [TestMethod]
+        public void ImporterWithoutAnyMappingFromGivenHeaderIndex()
+        {
+            // Arrange
+            CreateShiftedRowsWorkbook("Book1.xlsx", "Book6.xlsx", "TestClass", 4);
+            var stream = new FileStream("Book6.xlsx", FileMode.Open);
+
+
```
Comment (donnytian) [inline Npoi.Mapper/test/ImportGeneralTests.cs]: blank line

File: Npoi.Mapper/test/ExportGeneralTests.cs [hunk 9]
```diff
@@ -414,5 +414,86 @@ public void Map_WithIndexAndName_ShouldExportCustomColumnName()
...
+            // Prepare
+            const string existingFile = "Book4.xlsx";
+            const string sheetName = "Allocations";
+            const int headerRowIndex = 5;
+
+            if (File.Exists(existingFile)) File.Delete(existingFile);
+
+            CreateShiftedRowsWorkbook("Book1.xlsx", existingFile, "Allocations", headerRowIndex);
+
+            var exporter = new Mapper(existingFile) {HeaderRowIndex = headerRowIndex};
```
Comment (donnytian) [inline Npoi.Mapper/test/ExportGeneralTests.cs]: pls insert spaces:
```cs
var exporter = new Mapper(existingFile) { HeaderRowIndex = headerRowIndex };
                                         ^                               ^

```

**Comment by donnytian** [comment]
also we need change all places of usage of the first row.

Like in method `GetDynamicType`, seems you missed to update.

**Comment by waldimen** [comment]
> Hi, thanks for this PR!
> LGTM, can you please address some minor issues?

I resolved the issues, let me know if it is OK now

**Comment by Rwing** [comment]
Excuse me, is there any process?

File: Npoi.Mapper/src/Npoi.Mapper/Mapper.cs [hunk 10]
```diff
@@ -862,16 +864,17 @@ private void LoadWorkbookFromFile(string path)
         private void Put<T>(ISheet sheet, IEnumerable<T> objects, bool overwrite)
         {
             var sheetName = sheet.SheetName;
-            var firstRow = sheet.GetRow(sheet.FirstRowNum);
+            int firstRowIndex = HeaderRowIndex < 0 ? sheet.FirstRowNum : HeaderRowIndex;
```
Comment (Rwing) [inline Npoi.Mapper/src/Npoi.Mapper/Mapper.cs]: use 'var' is better

**Comment by donnytian** [comment]
replace it with `FirstRowIndex`, see #56

**Pull request closed without merging**